### PR TITLE
Add ability to probe specified sensor on any axis

### PR DESCRIPTION
**Pull request by StephanCoetzee**
This PR allows us to probe any sensor on any axis is any direction with the serial command:
$A{AXIS},{SENSOR},{DIRECTION}

For example, the command
$AY,0,1

Will move axis Y in the negative direction until sensor 0 is reached. Right now, we only have one probe sensor, but this code lays the groundwork for adding more sensors. Hard limits are also enabled when an axis is moving.

In a future PR, the current magazine monitor probing code will be replaced to make use of this code.

### CODE REVIEW

File: limits.c [hunk 1]
```diff
@@ -116,25 +116,18 @@ void limits_update_homing_values(uint8_t cycle_mask, float * homing_rate, float
...
-      if ((flipped & (1 << idx)) ^ approach) {
-        target[idx] = -travel;
-      } else {
-        target[idx] = travel;
-      }
-    } else {
-      target[idx] = 0;
-    }
+    const int8_t mult = (((flipped & (1 << idx)) ^ approach) ? -1 : 1);
+    target[idx] = (bit_istrue(cycle_mask, bit(idx))) ? (mult*travel) : 0;
```
Comment (Jeff-Ciesielski) [inline limits.c]: In a stance I don't usually take, I think that you might be a little better off with fewer braces here.  The outer pair on mult can probably go, as can the braces around the bit_istrue call.

Also: whitespace around `mul*travel`

File: nuts_bolts.h [hunk 2]
```diff
@@ -26,10 +26,12 @@
...
 #define N_AXIS 4 // Number of axes
-#define X_AXIS 0 // Axis indexing value. Must start with 0 and be continuous.
-#define Y_AXIS 1
-#define Z_AXIS 2
-#define C_AXIS 3
+enum e_axis {
+  X_AXIS = 0,
+  Y_AXIS,
+  Z_AXIS,
+  C_AXIS
```
Comment (Jeff-Ciesielski) [inline nuts_bolts.h]: You could actually make n_axis the last member of this enum.  That way if you ever expand, you don't have to worry about updating two things.

File: probe.c [hunk 3]
```diff
@@ -21,13 +21,136 @@
...
+  volatile uint8_t *in_port;
+};
+
+// Idx, Mask, PIN - Note PIN is not the
+// pin of the microcontroller, but Port In.
+// The indexing in this map should
+// reflect the order in the probe_sensor enum
+struct sensor sensor_map[N_SENSORS] = {
+  {MAG_SENSOR, MAGAZINE_ALIGNMENT_MASK, &MAGAZINE_ALIGNMENT_PIN}
+};
```
Comment (Jeff-Ciesielski) [inline probe.c:45]: If this is static to the probe code, you can combine the declaration and instantiation into an anonymous struct:

```
static struct {
...
} sensor_map [] = {
  {
  ...
  },
}
```

Also, please use named initializers in lieu of positional initialization.

File: probe.c [hunk 4]
```diff
@@ -21,13 +21,136 @@
...
+  /*  sensor - Which sensor to home
+   *  axis - Which axis to move to the sensor
+   *  dir - true = neg; false = pos
+   */
+
+  // Set the active probe
+  probe.active_sensor = sensor;
+
+  // Approach has all bits set (neg dir) or none (pos dir)
+  uint8_t approach = dir ? ~0 : 0;
```
Comment (Jeff-Ciesielski) [inline probe.c]: const?

Comment (Jeff-Ciesielski) [reply probe.c]: Const all the things in this function that should be immutable.

File: probe.h [hunk 5]
```diff
@@ -17,23 +17,49 @@
...
 #define PROBE_ACTIVE  1 // Actively watching the input pin.
 
+struct probe_state {
+  uint8_t active_sensor;  // The currently active probe seonsor
+  uint8_t probe_reached;  // Flag to indicate if active probe is reached
+  uint8_t isprobing;
+
+};
+
+struct probe_state probe;
```
Comment (Jeff-Ciesielski) [inline probe.h]: Please don't instantiate data in headers.

File: probe.h [hunk 6]
```diff
@@ -17,23 +17,49 @@
...
+
+enum e_sensor {
+  MAG_SENSOR = 0,
+  KEY_SENSOR,
+  e_sensor_types
```
Comment (Jeff-Ciesielski) [inline probe.h]: All caps.

File: stepper.c [hunk 7]
```diff
@@ -359,14 +359,13 @@ void st_limit_check()
...
-  if (!limits.ishoming) {
-      request_report(REQUEST_STATUS_REPORT|REQUEST_LIMIT_REPORT,LINENUMBER_EMPTY_BLOCK);
-  } else { 
-    bit_true(sys.state, STATE_HOME_ADJUST);
+    // If an axis is done homing, clear the corresponding bit in limits.ishoming     
+    limits.ishoming &= ~(must_stop >> LIMIT_BIT_SHIFT);
 
-  }
+    if (!limits.ishoming)
+        request_report(REQUEST_STATUS_REPORT | REQUEST_LIMIT_REPORT, LINENUMBER_EMPTY_BLOCK);
```
Comment (Jeff-Ciesielski) [inline stepper.c]: Indentation seems off here.

File: stepper.c [hunk 8]
```diff
@@ -403,7 +402,7 @@ void st_force_check()
   const uint8_t positive_stop = (positive_direction && (FORCE_VAL >= limits.bump_grip_force));
   const uint8_t negative_stop = (!positive_direction && (FORCE_VAL <= limits.bump_grip_force));
     
-  if (positive_stop | negative_stop) {
+  if (positive_stop || negative_stop) {
```
Comment (Jeff-Ciesielski) [inline stepper.c:405]: I'm fairly confident that a binary works and (in a miniscule way) potentially the more efficient choice here.

File: system.c [hunk 9]
```diff
@@ -265,7 +266,35 @@ uint8_t system_execute_line(char *line)
           if ( line[++char_counter] != 0 ) { return(STATUS_INVALID_STATEMENT); }
           else { report_ngc_parameters(); }
           break;
-
+        case 'A': // Probe: Specify sensor, axis and specified directionX,
+          char_counter++;
+          uint8_t axis = line[char_counter];
+          axis = get_axis_idx(axis);
+          //Axis
+          if ((axis == N_AXIS) || (line[++char_counter] != ','))
```
Comment (Jeff-Ciesielski) [inline system.c]: Do any other gcode commands include commas?  I'd rather see it be idiomatic if possible

File: system.c [hunk 10]
```diff
@@ -265,7 +266,35 @@ uint8_t system_execute_line(char *line)
...
+          if (!read_float(line, &char_counter, &value))
+            return(STATUS_BAD_NUMBER_FORMAT);      
+          uint8_t dir = (uint8_t)value;
+          if (dir > 1)
+            return(STATUS_BAD_NUMBER_FORMAT);
+          probe_move_to_sensor(sensor, axis, dir);
+          // End of system command:
+          if (line[++char_counter] != 0)
+            return(STATUS_INVALID_STATEMENT);
+          break;
```
Comment (Jeff-Ciesielski) [inline system.c:297]: Could you toss some line breaks in between the logical blocks here (probably before every comment)? This glob is hard to parse visually.

File: system.h [hunk 11]
```diff
@@ -92,7 +90,7 @@
 // Define global system variables
 typedef struct {
   uint8_t abort;                 // System abort flag. Forces exit back to main loop for reset.
-  uint8_t state;                 // Tracks the current state of Grbl.
+  uint16_t state;                 // Tracks the current state of Grbl.
```
Comment (Jeff-Ciesielski) [inline system.h:93]: Can you do me a favor and grep around to make sure that we're not relying on an 8 bit state variable anywhere?  I'd hate for higher order bits to get truncated accidentally due to an eager cast.

File: stepper.c (hunk 8 above)
Comment (StephanCoetzee) [reply stepper.c:405]: Yes, it works (I added it in https://github.com/keyme/grbl/pull/59). However, from a logical standpoint, I think the logical OR makes more sense than the bitwise OR and is easier to understand for someone reviewing the code who is not too familiar with it, since we check two separate boolean values.

File: system.c (hunk 10 above)
Comment (StephanCoetzee) [reply system.c:297]: I intentionally did not add lines to keep the format consistent with the rest of this 236 line function. I think such stylistic changes should be made on the file as a whole in a separate PR.

File: stepper.c (hunk 8 above)
Comment (Jeff-Ciesielski) [reply stepper.c:405]: Fair argument.

File: system.c (hunk 10 above)
Comment (Jeff-Ciesielski) [reply system.c:297]: Hah, fair enough.

File: limits.c [hunk 12]
```diff
@@ -116,25 +116,18 @@ void limits_update_homing_values(uint8_t cycle_mask, float * homing_rate, float
...
-      if ((flipped & (1 << idx)) ^ approach) {
-        target[idx] = -travel;
-      } else {
-        target[idx] = travel;
-      }
-    } else {
-      target[idx] = 0;
-    }
+    const int8_t mult = ((flipped & (1 << idx)) ^ approach) ? -1 : 1;
+    target[idx] = bit_istrue(cycle_mask, bit(idx)) ? (mult * travel) : 0;
```
Comment (mdicicco) [inline limits.c]: For the record, I don't like these "?" style if/then statements.  I find them hard to read.

Did you change the logic here, or was this just a cosmetic change?

File: system.h (hunk 11 above)
Comment (StephanCoetzee) [reply system.h:93]: Looks fine.

File: limits.c (hunk 12 above)
Comment (StephanCoetzee) [reply limits.c]: It's just a cosmetic change, since I prefer them to unnecessarily big nested if-else statements, but I can change it back.

Comment (Jeff-Ciesielski) [reply limits.c]: I used to hate them as well, but they grew on me as a matter of keeping immutable state.  I think in this case it might not have been the best choice, but consider something a little more complex:

```
int i = 5;
string result = i % 2 == 0 ? "a"
              : i % 3 == 0 ? "b"
              : i % 5 == 0 ? "c"
              : i % 7 == 0 ? "d"
              : "e";
```

Which works great as a poor man's match operator.

Comment (mdicicco) [reply limits.c]: I don't feel strongly.  As the primary maintainers, feel free to use what you think is best for you... I just think if/then's are more readable

**Comment by StephanCoetzee** [comment]
I'm ready - if you are happy @Jeff-Ciesielski and @mdicicco, you can merge.
